src/parse_impl.rs

```diff
@@ -0,0 +1,167 @@
+use crate::parse_fn::try_consume_fn;
+use crate::parse_utils::{consume_attributes, consume_stuff_until, consume_vis_marker};
+use crate::types::{Constant, ImplMember, TyDefinition, ValueExpr};
+use crate::types_edition::GroupSpan;
+use crate::{Attribute, TyExpr, VisMarker};
+use proc_macro2::{Group, Ident, TokenTree};
+use std::iter::Peekable;
+
+type TokenIter = Peekable<proc_macro2::token_stream::IntoIter>;
+
+pub(crate) fn consume_for(tokens: &mut TokenIter) -> Option<Ident> {
+    if let Some(TokenTree::Ident(ident)) = tokens.peek() {
+        if ident == "for" {
+            let tk_for = ident.clone();
+            tokens.next();
+            return Some(tk_for);
+        }
+    }
+
+    None
+}
+
+pub(crate) fn consume_constant(
+    tokens: &mut TokenIter,
+    attributes: Vec<Attribute>,
+    vis_marker: Option<VisMarker>,
+) -> Constant {
+    let tk_const = match tokens.next() {
+        Some(TokenTree::Ident(ident)) if ident == "const" => ident,
+        _ => panic!("cannot parse constant"),
+    };
+
+    let name = match tokens.next() {
+        Some(TokenTree::Ident(ident)) => ident,
+        _ => panic!("cannot parse constant"),
+    };
+
+    let tk_colon = match tokens.next() {
+        Some(TokenTree::Punct(punct)) if punct.as_char() == ':' => punct,
+        _ => panic!("cannot parse constant"),
+    };
+
+    let ty_tokens = consume_stuff_until(
+        tokens,
+        |tt| matches!(tt, TokenTree::Punct(punct) if punct.as_char() == '='),
+        true,
+    );
+
+    let tk_equals = match tokens.next() {
+        Some(TokenTree::Punct(punct)) if punct.as_char() == '=' => punct,
+        _ => panic!("cannot parse constant"),
+    };
+
+    let value_tokens = consume_stuff_until(
+        tokens,
+        |tt| matches!(tt, TokenTree::Punct(punct) if punct.as_char() == ';'),
+        true,
+    );
+
+    let tk_semicolon = match tokens.next() {
+        Some(TokenTree::Punct(punct)) if punct.as_char() == ';' => punct,
+        _ => panic!("cannot parse constant"),
+    };
+
+    Constant {
+        attributes,
+        vis_marker,
+        tk_const,
+        name,
+        tk_colon,
+        ty: TyExpr { tokens: ty_tokens },
+        tk_equals,
+        initializer: ValueExpr {
+            tokens: value_tokens,
+        },
+        tk_semicolon,
+    }
+}
+
+pub(crate) fn consume_assoc_ty(
+    tokens: &mut TokenIter,
+    attributes: Vec<Attribute>,
+    vis_marker: Option<VisMarker>,
+) -> TyDefinition {
```

Comment:
This should be called `consume_type_definition`. If called from the top level, it might parse a type alias, not an associated type.

(Also, for consistency, it should arguably return Option)

Reply:
Renamed and optionalized 🙂